server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -103,82 +111,143 @@ public static ElasticsearchException validate(
+                        if (remoteUnauthorizedIndices != null) {
+                            for (var entry : remoteUnauthorizedIndices.entrySet()) {
+                                entry.setValue(entry.getValue().stream().filter(index -> index.endsWith(":" + resource) == false).toList());
+                            }
+                        }
                         break;
                     }
-                    if (unauthorizedException == null && exception instanceof ElasticsearchSecurityException securityException) {
-                        unauthorizedException = securityException;
+                    if (localSecurityException == null && remoteException instanceof ElasticsearchSecurityException securityException) {
+                        if (remoteAuthorizationExceptions == null) {
+                            remoteAuthorizationExceptions = new HashMap<>();
+                            remoteUnauthorizedIndices = new HashMap<>();
+                        }
+                        remoteAuthorizationExceptions.putIfAbsent(projectAlias, securityException);
+                        remoteUnauthorizedIndices.computeIfAbsent(projectAlias, k -> new ArrayList<>()).add(remoteExpression);
                     }
                 }
                 if (foundFlat) {
                     continue;
                 }
-                // Prefer reporting 403 over 404
-                if (unauthorizedException != null) {
-                    return unauthorizedException;
-                }
-                // Prefer reporting 404 on origin over linked projects
-                if (localException != null) {
+                if (localSecurityException != null) {
+                    if (localAuthorizationException == null) {
+                        localAuthorizationException = localSecurityException;
+                        localUnauthorizedIndices = new ArrayList<>();
+                    }
+                    localUnauthorizedIndices.add(originalExpression);
+                } else if (localException != null) {
                     assert localException instanceof IndexNotFoundException
                         : "Expected local exception to be IndexNotFoundException, but found: " + localException;
-                    return localException;
+                    notFoundException = (IndexNotFoundException) localException;
                 } else {
                     assert false == remoteExpressions.isEmpty() : "expected remote expressions to be non-empty";
-                    return new IndexNotFoundException(remoteExpressions.iterator().next());
+                    notFoundException = new IndexNotFoundException(remoteExpressions.iterator().next());
                 }
             }
         }
-        // if we didn't throw before it means that we can proceed with the request
-        return null;
+
+        if (localAuthorizationException == null && remoteAuthorizationExceptions == null) {
+            return notFoundException;
```

Comment:
This isn't quite right with the "add to maps then clear on foundFlat"  approach. `remoteAuthorizationExceptions` might be non-null but an empty map (or a map with stale entry with what you have right now). When this happens, we will ignore  `notFoundException` and go into the `else` branch which either finds nothing or a stale entry.